src/mam/mod_mam_cache_user.erl

```diff
@@ -106,103 +81,65 @@ muc_hooks(HostType) ->
-            undefined;
-        UserID ->
-            UserID
+            undefined
     end.
 
 -spec store_archive_id(mod_mam:archive_id(), mongooseim:host_type(), jid:jid())
         -> mod_mam:archive_id().
-store_archive_id(UserID, _HostType, ArcJID) ->
-    maybe_cache_archive_id(ArcJID, UserID),
-    UserID.
+store_archive_id(ArchId, HostType, ArcJid) ->
+    case erase(mam_not_cached_flag) of
+        undefined ->
+            ArchId;
+        true ->
+            CacheName = cache_name(HostType),
+            segmented_cache:merge_entry(CacheName, key(ArcJid), #{id => ArchId}),
+            ArchId
+    end.
 
 -spec remove_archive(Acc :: map(), HostType :: mongooseim:host_type(),
-                     UserID :: mod_mam:archive_id(), ArcJID :: jid:jid()) -> map().
-remove_archive(Acc, _HostType, _UserID, ArcJID) ->
-    clean_cache(ArcJID),
+                     ArchId :: mod_mam:archive_id(), ArcJid :: jid:jid()) -> map().
+remove_archive(Acc, HostType, _UserID, ArcJid) ->
+    CacheName = cache_name(HostType),
+    segmented_cache:delete_entry(CacheName, key(ArcJid)),
     Acc.
 
 %%====================================================================
-%% Internal functions
+%% internal
 %%====================================================================
-
--spec maybe_cache_archive_id(jid:jid(), mod_mam:archive_id()) -> ok.
-maybe_cache_archive_id(ArcJID, UserID) ->
-    case erase(mam_not_cached_flag) of
-        undefined ->
-            ok;
-        true ->
-            cache_archive_id(ArcJID, UserID)
+-spec cache_name(mongooseim:host_type()) -> atom().
+cache_name(HostType) ->
+    case gen_mod:get_module_opt(HostType, ?MODULE, cache_module, internal) of
+        internal -> cache_name(HostType, ?MODULE);
+        CacheMod -> cache_name(HostType, CacheMod)
     end.
 
-%% @doc Put the user id into cache.
-%% @private
--spec cache_archive_id(jid:jid(), mod_mam:archive_id()) -> ok.
-cache_archive_id(ArcJID, UserID) ->
-    gen_server:call(srv_name(), {cache_archive_id, ArcJID, UserID}).
-
--spec lookup_archive_id(jid:jid()) -> mod_mam:archive_id() | not_found.
-lookup_archive_id(ArcJID) ->
-    try
-        ets:lookup_element(tbl_name_archive_id(), su_key(ArcJID), 2)
-    catch error:badarg ->
-        not_found
-    end.
+-spec cache_name(mongooseim:host_type(), module()) -> atom().
+cache_name(HostType, Module) ->
+    gen_mod:get_module_proc(HostType, Module).
 
--spec clean_cache(jid:jid()) -> ok.
-clean_cache(ArcJID) ->
-    %% Send a broadcast message.
-    case pg:get_members(group_name()) of
-        [_|_] = Pids ->
-            [gen_server:cast(Pid, {remove_user, ArcJID})
-            || Pid <- Pids],
-            ok;
-        _ ->
-            ok
+-spec start_cache(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start_cache(HostType, Opts) ->
+    case gen_mod:get_opt(cache_module, Opts, internal) of
+        internal -> mod_cache_users:start_new_cache(HostType, ?MODULE, Opts);
```

Comment:
Just a note from me: I think it's better not to introduce such coupling between `mod_*` modules - for example I spent quite a lot of time decoupling `mod_disco` from other modules and putting the shared functionality in a separate module called `mongoose_disco`. So here I would see common functionality in something like `mongoose_user_cache`.